### PR TITLE
Ability to search in `SingleChoiceParameterType`  drop down lists

### PR DESCRIPTION
**Pull request by Auriga2**
### Description of the Change
Users can now type on the SingleChoiceParameterType drop downs, which would filter the items in the drop down list.

### Alternate Designs
Tried a few different options but using `controlsfx.controls.SearchableComboBox` seems to handle it very well.

### Why Should This Be In Core?
Better ux, specially with longer drop down lists.

### Benefits
Faster ux, specially with longer drop down lists.

### Possible Drawbacks
n/a

### Verification Process
Start typing in a `SingleChoiceParameterType` drop down. It should filter options based on what you have typed.
E.g.: `Mode` and `Type Category` drop downs in `Select Top N` plugin
Drop downs in `Extract words from Text`, `Merge Nodes`, `Merge Transactions` plugins etc.

### Applicable Issues
https://github.com/constellation-app/constellation/issues/347

### CODE REVIEW

File: CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/parameters/types/SingleChoiceParameterType.java [hunk 1]
```diff
@@ -128,9 +128,22 @@
...
+
+            //Clear the existing selection
+            parameter.setObjectValue(null);
+
         parameterValue.setOptions(options);
         parameter.setProperty(CHOICES, new Object());
+        }
+    }
+
+    private static boolean optionsChanged(PluginParameter<SingleChoiceParameterValue> parameter, List<String> options) {
```
Comment (antares1470) [inline CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/parameters/types/SingleChoiceParameterType.java:143]: both parameters can be made final

File: CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/parameters/types/SingleChoiceParameterType.java [hunk 2]
```diff
@@ -128,9 +128,22 @@
...
+        final SingleChoiceParameterValue parameterValue = parameter.getParameterValue();
+        boolean bol = !options.equals(parameterValue.getOptions());
+        return bol;
```
Comment (antares1470) [inline CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/parameters/types/SingleChoiceParameterType.java:146]: these lines can be reduced to `return !options.equals(parameter.getParameterValue().getOptions())`

**Review by antares1470 — CHANGES_REQUESTED** [review]
If you're able to add a whatsnew entry as well that would be great

File: CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/parameters/types/SingleChoiceParameterType.java (hunk 2 above)
Comment (Auriga2) [reply CorePluginFramework/src/au/gov/asd/tac/constellation/plugins/parameters/types/SingleChoiceParameterType.java:146]: Yeah already fixed it but didn't push that file. Added that for testing.

**Review by arcturus2 — CHANGES_REQUESTED** [review]
Thank you @Auriga2 this is really nice to have as a feature and works well. There seems to be one issue with a delayed update to sub lists and I had to click twice to get the values for the list.

I noticed that if I start to type after the first click, items do show up so it looks like the list is populated but its not visible until the second click/GUI update.

Do you think this could be fixed and if its an easy fix?

drop down list delay

**Comment by Auriga2** [comment]
@arcturus2 I can recreate it only when running after a clean build. Could you recreate it without that?

**Comment by Auriga2** [comment]
No need to clean and build to recreate. But still need to close the consty and debug run each time. This is a bit tricky as I could not able to recreate in ControlsFX yet. I tried to fix locally by initializing some values in the drop downs by default. 

If I set to Node (the longer list), selecting the transactions throws the same `NoSuchElementException` exception as before. 
If I set to Transaction ( the shorter list),  selecting the Node doesn't show the entire list or the side scroll bar until clicked twice on it. Looks like this is still initialized to the smaller size still.

Also tried to select an option by default. This populated the selection, but the behavior in the drop down list is still the same.

This issue presents only when the DAV is loaded for the first time , and also only on a dependent drop down, probably only in the `Select Top N` plugin. (Are there any other such drop downs in Consty?)

**Comment by Auriga2** [comment]
Found a workaround. Also removed some redundant code.

**Review by arcturus2 — APPROVED** [review]
Your a legend @Auriga2 awesome stuff and it works really well.

File: CoreWhatsNewView/src/au/gov/asd/tac/constellation/views/whatsnew/whatsnew.txt [hunk 3]
```diff
@@ -1,6 +1,9 @@
 == 3030-12-31 Getting Started
 <p>If you're new to CONSTELLATION, read the <a href="" helpId="au.gov.asd.tac.constellation.functionality.introduction">introduction</a>.</p>
 
+== 2021-04-01 Ability to Search in Single Choice Drop-Down Lists
+<p>Single Choice Drop-Downs are now editable, which would filter the options in the drop down list based on what is typed.</p>
```
Comment (antares1470) [inline CoreWhatsNewView/src/au/gov/asd/tac/constellation/views/whatsnew/whatsnew.txt:5]: change "editable" to "searchable" and "which would" to "which allows you to"

**Review by antares1470 — CHANGES_REQUESTED** [review]
Tested and looks good otherwise

File: CoreWhatsNewView/src/au/gov/asd/tac/constellation/views/whatsnew/whatsnew.txt (hunk 3 above)
Comment (Auriga2) [reply CoreWhatsNewView/src/au/gov/asd/tac/constellation/views/whatsnew/whatsnew.txt]: Yes they are not editable. Fixed.

**Review by antares1470 — APPROVED** [review]

**Pull request merged**